### ISSUE TITLE
Create NREL-hosted instance of OpenPATH for ease of use by public agencies

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#721]
@aguttman As we discussed, the plan is to have multiple enclaves, one for each program or study.
- each enclave would contain multiple inter-connected containers and be accessible via reverse proxy
- we would also create a study-specific landing page, which we would describe the study and ask users to install OpenPATH and then scan a QR code to join the study

**Comment by shankari** [comment]
@aguttman

The first step for this will be to create one enclave for internal NREL use.
Creating this enclave will require permission from multiple NREL departments:
- cyber security
- legal
- contracting
- communications
etc

Since this is a complicated endeavor we will use an internal NREL project as an example to get started. @shankari will continue coordinating this since she has the context.
- Ask IT to create staging and production environments consisting of one webapp container and two scheduled tasks similar to https://github.com/e-mission/e-mission-docker/tree/master/examples/em-server-multi-tier-cronjob. The two scheduled tasks are the ones in the current crontab in the example.
   - This has been approved by cyber already so it just needs to be reinstated after a security scan.
-  NREL IT will configure https://openpath-stage.nrel.gov/study/<study_name>/api and https://openpath.nrel.gov/study/<study_name>/api to point to the webapps in these two environments (ticket pending). 
- work with comms to create a static page describing the NREL internal project and hosting it at something like https://nrel.gov/openpath/<study_name>
- publish the NREL apps to the stores after legal permission (@agutmann should get familiar with this)

In parallel, we should start preparing for the more complex solution above. The immediate tasks that I can list to prepare include:
- we need to fix the compatibility issues between mongodb and documentDB (https://github.com/e-mission/e-mission-docs/issues/597) without hacks like https://github.com/e-mission/e-mission-docker/pull/17 or other read-after-write issues
- we need to store the created models in the database instead of in files
- we need to figure out the authentication solution for the deployer dashboard, which is written in R Shiny and get it approved by cyber and cloud services. cyber prefers AWS Cognito, but we can also use a simpler approach since we will only have a few admins for each program.

**Comment by shankari** [comment]
This should fix https://github.com/e-mission/e-mission-docs/issues/326 and provide a template for fixing https://github.com/e-mission/e-mission-docs/issues/452

**Comment by aGuttman** [comment]
Of these, I am first focusing on compatibility issues between mongodb and documentDB. I will familiarize myself with both to the extent needed to understand the problems present and what a proper solution should be like.

Secondarily, I will be attentive of @shankari 's communication and progress on creating the internal NREL enclave so that I can understand the process and be able to take point on future enclave creation.

**Comment by shankari** [comment]
Another, more serious compatibility issue: https://github.nrel.gov/nrel-cloud-computing/emissionlhd/issues/10#issuecomment-38430

**Comment by shankari** [comment]
I have currently worked around this on the NREL setup by adding a sleep to every DB call, but that is so bad and hacky that I will never merge it to master (https://github.nrel.gov/nrel-cloud-computing/emissionlhd/pull/16)

**Comment by shankari** [comment]
As pointed out by @jgu2 here are the pages on DocumentDB related to the issue:

https://docs.aws.amazon.com/documentdb/latest/developerguide/how-it-works.html#how-it-works.replication
https://docs.aws.amazon.com/documentdb/latest/developerguide/transactions.html#transactions-isolation-level
https://docs.aws.amazon.com/documentdb/latest/developerguide/how-it-works.html#durability-consistency-isolation.read-consistency

Ddocdb cluster is distributed, it has primary node and replicas, primary node supports read/write, and replica supports read only. The reads from a read replica are eventually consistent, often less than 100ms replica lag. To keep read-after-write consistency, it can setup the read preference,

```
db.example.find().readPref('primary')
or 
db.example.find().readPref('primaryPreferred')
```

Since we should be using the timeseries interface (`emission.storage.timeseries.abstract_timeseries`) for most calls, there should not be a lot of calls to find(). So maybe we should start with this fix and see if it works?

**Comment by shankari** [comment]
Although the real fix should be to look at the control flow and see where we have read-after-write dependencies.
Here are all the write calls:

```
$ egrep -R -l "insert_one|replace_one|update_one" emission | grep -v tests | grep -v .pyc
```

which currently returns 

```
emission/net/usercache/builtin_usercache.py
emission/net/ext_service/push/notify_interface_impl/firebase.py
emission/net/api/usercache.py
emission/core/wrapper/user.py
emission/core/wrapper/client.py
emission/core/get_database.py
emission/storage/timeseries/builtin_timeseries.py
```

**Comment by shankari** [comment]
Tasks for the documentDB support:
- Talk to cyber and get access to the documentDB staging environment
- Read the current documentDB support for indices and see whether HASHED and GEOSPHERE are now supported
   - If yes, remove the hack to remove those unsupported indices
   - If still not supported:
      - run the unit test suite against a mongodb on your laptop
      - run the unit test suite against the documentDB staging environment
      - if documentDB is comparable to mongodb on laptop (no more than 10% slower), we are revisit in 6 months
      - if documentDB is significantly slower than mongodb on laptop, instead of removing index, replace with HASHED index types that are supported (e.g. ASCENDING)
         - GEOSPHERE cannot be replaced by ascending/descending, but it is not used extensively just yet
- for the timing issue, revise "read replicas" and the read-after-write problem
   - currently hacked by "sleep" to every DB call
   - cyber suggested using a `readPref` (https://github.com/e-mission/e-mission-docs/issues/721#issuecomment-1109078184)
   - change all DB read calls to use the read preference - see if all the tests pass
      - if they do, commit the change and generate a pull request to master
      - if they don't, then discuss further

**Comment by shankari** [comment]
### Tasks for May 3rd @aGuttman 

- [ ] Request access to OpenPATH staging database 
- [ ] Get docker installed on your laptop
- [ ] Figure out how to run all the tests
- [ ] Investigate current status of DocumentDB indices
- [ ] **stretch goal** if they do now support HASHED and GEOSPHERE, remove the SIMPLE_INDEX hack from the docker file linked above and submit PR

**Comment by shankari** [comment]
We also need to have a demographic survey for each user. We have typically used external surveys (Qualtrics/Google Forms) before, but those have the following limitations:
- they are not approved survey tools at NREL
- they are not integrated with the rest of the data, so we can't ensure that we only popup the survey if the user has not yet answered it earlier
- we also can't experiment with fancy solutions where we break up the onboarding survey and ask the questions in little dribbles

A solution that would address all those issues is to store the survey information in mongodb, just like everything else. But we don't want to create a survey builder. Instead, we will use kobotoolbox to create a survey which we will display using the enketo library.

The UNSW group has already integrated with enketo core in the https://github.com/e-mission/e-mission-phone/tree/rciti branch, so let's start by exploring that approach.

This feature is tracked in https://github.com/e-mission/e-mission-docs/issues/727

**Comment by shankari** [comment]
From @jgu2:

> It looks like you are using a hard-coded db name in code of emission called  “Stage_database”. For best practice, it’s recommended to use different db for different deploy environment, as we are moving towards production deploy, a production db will be created on the same docdb cluster.
 
> What’s more, as OpenPath needs multi-containers for different centers/organizations, ideally each center/organization would have its own db setup.
 
> We can setup DB name in an environment variable, for example DB_NAME, where you can get access to according db via the variable. If this way, you will need to change your code to adapt it. Otherwise, all your data of stage/production from different centers/orgs in the future would be written into same db. What’s your thoughts?

From @shankari:

> Makes sense.

> For the record, the reason that the DB/collection name is hardcoded is because I typically run multiple different MongoDB containers for the different instances. Too bad that isn’t an option on the AWS cluster, but we can adapt it pretty easily, I think.

**Comment by shankari** [comment]
@aGuttman will have the primary DB changes done by tomorrow:
- [ ] Fix indices
- [ ] Deal with read replicas and readPreferred
- [ ] Configure DB name

Then he will work on #724

**Comment by aGuttman** [comment]
> ### Tasks for May 3rd @aGuttman
> * [x]  Request access to OpenPATH staging database
> * [x]  Get docker installed on your laptop
> * [x]  Figure out how to run all the tests
> * [x]  Investigate current status of DocumentDB indices
> * [ ]  **stretch goal** if they do now support HASHED and GEOSPHERE, remove the SIMPLE_INDEX hack from the docker file linked above and submit PR

**Comment by shankari** [comment]
> Not sure if this is the best way to do it, but thought it was attractive after seeing the URL solution to read preference to keep all these considerations in a single part of the already used db.config file.

Agree that this is attractive. What are some additional design options, and can you list the pros and cons before choosing this one?

**Comment by shankari** [comment]
wrt this list, here's my initial pass-through:

```
    ./emission/core/__pycache__/get_database.cpython-37.pyc : compile artifact, ignore
    ./emission/core/get_database.py  : already identified
    ./emission/analysis/classification/inference/Classifier.ipynb : remove, fairly obsolete
    **./emission/integrationTests/storageTests/TestMongodbAuth.py** : this is a test of the authenticated connections to mongodb. Use it to test the solution we come up with
    ** ./emission/storage/timeseries/abstract_timeseries.py**: Look at the history/blame for the code which includes it, and see if it can be rewritten to use get_database.
    **./emission/storage/timeseries/__pycache__/abstract_timeseries.cpython-37.pyc**: compile artifact, ignore
    **./bin/historical/migrations/rename_client_ts_to_ts.py**: historical script, ignore. Maybe add a readme to the directory indicating that these are for reference and are not guaranteed to work now.
    **./bin/plot_trip.py**: Pretty sure this is obsolete. Check history and remove.
    **./setup/db_auth.py**: This is a script to setup the mongodb auth properly based on username and password. adapt to handle the DB name as well.
```

**Comment by aGuttman** [comment]
> @aGuttman
> 
> > URL when connecting to remote documentDB instance can include a read preference setting. Setting this to primaryPreferred should avoid read-after-write problem without changing every find() call in code to find().readPref('primaryPreferred').
> 
> can you confirm that you tested this? the simple test-after-set function above should be sufficient

I did test this with the `test_save_and_list()` example typed out above and it did work, though I'm confused about how to interprate the behavior. `primary`, `primaryPreferred` and `secondaryPreferred` all produced the correct output. `secondary` times out. Invalid settings cause a warning, but the correct output is still produced. In my testing I haven't been able to get an empty list back like the example. I'm not including any sleeps.

Using primary read preference in the URL to avoid read-after-write problems is definitely an intended feature, as seen in the first example under _Multiple Connection Pools_ here: https://docs.aws.amazon.com/documentdb/latest/developerguide/connect-to-replica-set.html

Also, if we don't take advantage of replica sets, it seems like we could just leave out the rs0 part of the URL.

> If you omit `/?replicaSet=rs0`, the client routes all requests to the cluster endpoint, that is, your primary instance

**Comment by shankari** [comment]
> In my testing I haven't been able to get an empty list back like the example. I'm not including any sleeps.

That might be because your connection to the database is slow so it acts as an implicit sleep. That is also probably why the tests take a long time to run. I think we can rely on the documentation for now (maybe leave out the `rs0`) and revisit if we still have issues.

**Comment by aGuttman** [comment]
> @aGuttman we don't store anything outside that database, so do you even need to parse it out? Maybe if you pass in a URL with the database embedded, you don't need to specify the database any more. From the SO post that you linked, it looks like maybe something like the following would just work
> 
> ```
> MongoClient("mongodb://<user>:<password>@documentdb-moderate-1.<rest_of_host>:27017/<dbName>?ssl=true&ssl_ca_certs=<path_to_certs>/rds-combined-ca-bundle.pem&replicaSet=rs0&readPreference=primaryPreferred&retryWrites=false").Stage_uuids.count_documents()
> ```
> 
> you could then replace
> 
> ```
> print("Connecting to database URL "+url)
> _current_db = MongoClient(url).Stage_database
> ```
> 
> with
> 
> ```
> print("Connecting to database URL "+url)
> _current_db = MongoClient(url)
> ```

I don't think this works. I'm trying to do something like it now  and it's returning errors like `TypeError: 'Collection' object is not callable. If you meant to call the 'delete_many' method on a 'Database' object it is failing because no such method exists.` I might be misunderstanding something, though I am able to get it to work parsing out the name and using dictionary style access.

```
url = config_data["timeseries"]["url"]
parsed=pymongo.uri_parser.parse_uri(url)
...
print("Connecting to database URL "+url)
_current_db = MongoClient(url)[parsed['database']]
```

It would be nice to avoid more parsing than necessary, but I think the name does need to be specified on the client object (?) in order to use it. Given that, it needs to be passed somehow from an external source. We can make it its own field in a json, but that just changes where the parse happens. Maybe the pymongo parser is slower, but we only need it once at the start and it just feels cleaner to me to specify it all in the url instead of adding another field to the config (or making an additional config file). Those are the only options I'm able to think of right now, and of those, I just like the url option better, even at the cost of using a potentially slightly less effective parser.

**Comment by aGuttman** [comment]
Also let's hold off on my pull request for a minute. `tests/netTests/TestBuiltinUserCacheHandlerInput.py` is failing. I can go back on all my changes expect the ones that need to be there (swap ASCENDING for HASHED, remove GEOSPHERE entirely or just remove sparse=True) but `testMoveDuplicateKey` and `testTwoLongTermCalls` fail. I'm going to look at what those two do.

**Comment by shankari** [comment]
If you see https://github.nrel.gov/nrel-cloud-computing/emissionlhd/issues/10#issuecomment-38429, laptop to DocumentDB always worked with the read preference. 

> on the laptop, the communication overhead for the call is high, which effectively functions as a sleep

You need to get @jgu2 to create an AWS EC2 instance (maybe to ssh into the staging instance?) and then run the script from there.

**Comment by shankari** [comment]
> Given that, it needs to be passed somehow from an external source. We can make it its own field in a json, but that just changes where the parse happens. Maybe the pymongo parser is slower, but we only need it once at the start and it just feels cleaner to me to specify it all in the url instead of adding another field to the config (or making an additional config file). Those are the only options I'm able to think of right now, and of those, I just like the url option better, even at the cost of using a potentially slightly less effective parser.

This makes sense to me. Please submit a PR for this as well. You probably want to communicate the new URL format (after it is finalized) with @jgu2 so he can set the environment variables correctly.

**Comment by shankari** [comment]
- [ ] Get access to an AWS EC2 instance to test out the read preference
- [ ] Debug and fix the test failures on DocumentDB
- [ ] Clean up if needed and submit PR for DB name

**Comment by aGuttman** [comment]
diffing the terminal output of both local and remote runs of `TestBuiltinUserCacheHandlerInput.py`just seems like nonsense to me because there is too much going on.
[diffTest.txt](https://github.com/e-mission/e-mission-docs/files/8665387/diffTest.txt)

**Comment by shankari** [comment]
every line is different because the timestamps are different. If you `grep -v` all the lines related to `Updated result for` and `cut -d ":" -f ...` to remove all the timestamp stuff, you should be able to see exactly what is different

**Comment by aGuttman** [comment]
> Ran 237 tests in 39327.855s
> 
> FAILED (failures=2, errors=15)

Just about 11 hours to run the tests remotely this time. No sleeping this time. I guess the timer from the last run didn't count that time either.

**Comment by aGuttman** [comment]
For the `TestBuiltinUserCacheHandlerInput.py` test failures:

- It appears to be a timing issue. Different individual tests in the file will fail from run to run, sometimes even at different places.
- Adding sleeps at certain points seems to help some failures, but not always, and often getting one test to pass is coupled with a new one failing.
- It does not appear to be an issue with replicas/read preference. Changing these in the url at least hasn't corresponded with more or less success.
- These failures do not seem to be a result of the index changes which is the only code change being tested. Commenting out/replacing the index changes don't affect the test.
- The failure is almost always an assert statement enforcing that a user cache is empty after calling `moveToLongTerm()` from `emission/net/usercache/builtin_usercache_handler.py`
- ie `AssertionError: 30 != 0` on 
```
# Then we move entries for user1 into longterm
enuah.UserCacheHandler.getUserCacheHandler(self.testUserUUID1).moveToLongTerm()

# Now, we have two sets of entries, so we will have 60 entries in longterm
self.assertEqual(len(self.uc1.getMessage()), 0)
```
Again, adding sleeps around this call seems to help sometimes, but not always.

Looking at the definition of `moveToLongTerm()` and seeing that the problem is that the cache is failing to be emptied, it seems like the problem would be in its call to `clearProcessedMessages()` in `emission/net/usercache/builtin_usercache.py` but that just boils down to
```
def clearProcessedMessages(self, timeQuery, key_list=None):
        del_query = self._get_msg_query(key_list, timeQuery)
        logging.debug("About to delete messages matching query %s" % del_query)
        del_result = self.db.delete_many(del_query)
        logging.debug("Delete result = %s" % del_result)
```
There doesn't seem to very much online discussion of documentDB issues online in general and I'm not finding much about synchronization issues with `delete_many()`, but I don't know. I'm not sure what to do from here.

**Comment by shankari** [comment]
@aGuttman I would:
1. document what you did with the sleeps and the amount of the sleep and document the failures that you encountered
2. try my prior sleep hack and see if that fixes it

Note that if we have `moveToLongTerm()` followed by an assertion to 0 but the actual value is 30, this could also just be a read of stale information.

**Comment by shankari** [comment]
The next thing I would try is to come up with an MRE that doesn't use any of the openpath code, but just pymongo directly.
For example, something like:

```
test_db = MongoClient(url).Test_inconsistency_database
test_coll = test_db.Test_collection
lengths = [10,100,1000,10000]
for l in lengths:
     # Insert l entries for l = 10, 100...
     for i in range(l):
         test_coll.insert({"i": i})
     # busy loop until we get l documents
     for test_coll.count_documents() < l:
         # potentially add sleep here
         print(test_coll.count_documents())
     # Delete all the entries
     test_coll.remove_many({})
     # busy loop until we get 0 documents
     for test_coll.count_documents() > 0:
         # potentially add sleep here
         print(test_coll.count_documents())    
```

**Comment by aGuttman** [comment]
I believe I've pinpointed where the error is happening, but I don't understand it and am not able to reproduce it outside of the e-mission code.

By adding logs, I can see that `moveToLongTerm()` (in `emission/net/usercache/builtin_usercache_handler.py`) consistently sees the correct number of entries. After processing them, it passes the same query that it uses to correctly find these entries into `clearProcessedMessages()` (in `emission/net/usercache/builtin_usercache.py`). `clearProcessedMessages()` uses these arguments to call `_get_msg_query()`, but when this query is used, it sometimes fails to find any entries.

The odd part is that `moveToLongTerm()` uses `getMessage()` (in `emission/net/usercache/builtin_usercache.py`) with arguments `None` and `time_query` to create `curr_entry_it` which it iterates through to find each entry. `getMessage()` simply calls `_get_msg_query()` (in `emission/net/usercache/builtin_usercache.py`) with `None` and `time_query` and returns the result as a list. `clearProcessedMessages()`calls `_get_msg_query()` with `None` and `time_query`, the exact same call and arguments that perviously worked, but uses the result directly as a filter to `delete_many()`. When used this way, `delete_many()` will often fail to find any entries and delete nothing.

I'm trying to get the entries that `clearProcessedMessages()` should have as a list and delete them iteratively since that seems to work for `moveToLongTerm()`, for some reason. I can get the right number of entires now but at the moment I'm a bit fried and can't properly think through deleting them right for some reason, though it shouldn't be too hard. That said I have no idea why it would work this way as opposed to the other.

**Comment by aGuttman** [comment]
```
emission/net/usercache/builtin_usercache_handler.py
99
-logging.debug("moveToLongTerm: Deleting all entries for query %s" % time_query)
+logging.debug("moveToLongTerm: Deleting all %d entries for query %s" % (len(unified_entry_list), time_query))
```

```
emission/net/usercache/builtin_usercache.py
153
+del_list = self.getMessage(key_list, timeQuery)
+logging.debug("clearProcessedMessages: itr: deleting %d entries" % len(del_list))
+logging.debug("clearProcessedMessages: deleting %d entries" % self.db.count_documents(del_query))
```

```
log
2022-05-12 16:20:04,089:DEBUG:4542789120:moveToLongTerm: Deleting all 30 entries for query <emission.storage.timeseries.timequery.TimeQuery object at 0x7fba41101910>
...
2022-05-12 16:20:04,140:DEBUG:4542789120:clearProcessedMessages: deleting 0 entries
2022-05-12 16:20:04,191:DEBUG:4542789120:clearProcessedMessages: deleting 0 entries
```

**Comment by aGuttman** [comment]
emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py

**Comment by aGuttman** [comment]
Note that wrapping the calls to moveToLongTerm() in a loop such as below does work:
```
        # Then, we move entries for the ios user into longterm
        count = 0
        while True:
            enuah.UserCacheHandler.getUserCacheHandler(self.testUserUUIDios).moveToLongTerm()
            if len(self.ucios.getMessage()) == 0:
                break
            count += 1
            logging.debug("move fail count: %d" % count)
        
        self.assertEqual(len(self.ucios.getMessage()), 0)
```

Doing this we can consistently pass the tests, but I've seen the number of retries per failure range from 3 to >300. Median is about 60 from my experience.

**Comment by aGuttman** [comment]
Setting read preference at database creation such as below had no effect:
```
from pymongo import MongoClient, ReadPreference

_current_db = MongoClient(url, read_preference=ReadPreference.PRIMARY).Stage_database
```

The per call setting suggested by amazon's documentation does not seem to be supported by pymongo:
```
db.example.find().readPref('primary')
```
This causes errors, even when trying various imports and I could not find usage online.

**Comment by aGuttman** [comment]
I'm seeing a lot more failures like this today, as opposed to the != 0 fails that I was focused on yesterday. I wonder if the timing conditions of working at home vs in office have an effect.
```
======================================================================
FAIL: testMoveToLongTerm (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 123, in testMoveToLongTerm
    self.assertEqual(edb.get_timeseries_db().estimated_document_count(), 60)
AssertionError: 62 != 60

======================================================================
FAIL: testTwoLongTermCalls (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 329, in testTwoLongTermCalls
    self.assertEqual(edb.get_timeseries_db().estimated_document_count(), 120)
AssertionError: 122 != 120

----------------------------------------------------------------------
Ran 4 tests in 222.911s

FAILED (failures=2)
```

**Comment by shankari** [comment]
On the other hand, I am in CA, I just re-ran the tests and got

```
======================================================================
FAIL: testMoveDuplicateKey (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 161, in testMoveDuplicateKey
    self.assertEqual(len(self.uc1.getMessage()), 30)
AssertionError: 0 != 30

======================================================================
FAIL: testMoveToLongTerm (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 87, in testMoveToLongTerm
    self.assertEqual(len(self.uc1.getMessage()), 30)
AssertionError: 0 != 30

======================================================================
FAIL: testMoveWhenEmpty (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 130, in testMoveWhenEmpty
    self.assertEqual(len(self.uc1.getMessage()), 0)
AssertionError: 30 != 0

======================================================================
FAIL: testTwoLongTermCalls (__main__.TestBuiltinUserCacheHandlerInput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py", line 259, in testTwoLongTermCalls
    self.assertEqual(len(self.uc1.getMessage()), 0)
AssertionError: 30 != 0

----------------------------------------------------------------------
Ran 4 tests in 209.991s

FAILED (failures=4)
```

**Comment by shankari** [comment]
@aGuttman If we are both going to test at the same time, we probably want to work on separate databases. If you have the dbname fix, could you submit it as well? Rebase the PR to something other than master, I will pull from it, and we can pick different database names.

**Comment by aGuttman** [comment]
Parsing the db name out of the config url works but I was light on my testing when I put it together the other day. Using the name "Stage_databse" in the url, everything gets parsed fine and I can connect and run, but using any other name gives an auth error.
```
pymongo.errors.OperationFailure: Authorization failure, full error: {'ok': 0.0, 'operationTime': Timestamp(1652462420, 1), 'code': 13, 'errmsg': 'Authorization failure'}
```

I believe this is a problem with the server or credentialing, not the e-missions code/ setup, but I don't understand enough to know. Even manually changing every instance of 'Stage_database' to a new name and rerunning the setup doesn't allow you to connect to a database with a different name.

ie:
find/replace "Stage_database" with "NewName" across the whole project
rerun `source setup/setup.sh`
run `source setup/activate.sh`
create unrelated file `tested.py` with `test_db = MongoClient(url).NewName`
`test_db.test_coll.insert({"a": 1})`
yields
`pymongo.errors.OperationFailure: Authorization failure, full error: {'ok': 0, 'operationTime': Timestamp(1652464086, 1), 'code': 13, 'errmsg': 'Authorization failure'}`

Doing the same with
`test_db = MongoClient(url).Stage_database`
works

I will email Jianli about it, but please let me know if it looks like I'm misunderstanding something.

**Comment by shankari** [comment]
@aGuttman i think that is because we have access only to the `Stage_database`; accessing a different database will likely require a different username and password. can you get @jgu2 to create two dev databases (one for each of us) and give us the credentials to connect to them

**Comment by aGuttman** [comment]
I'm noticing this from logging time_query, but I don't know what to make of it:
Here, we fail to find entries to delete:
```
2022-05-13 10:57:41,066:DEBUG:4712445440:moveToLongTerm: Deleting all 30 entries for query <emission.storage.timeseries.timequery.TimeQuery object at 0x7fc348145610>
2022-05-13 10:57:41,066:DEBUG:4712445440:time_query: call clear with: type: metadata.write_ts start: 1652465.999 end: 1652460.999
2022-05-13 10:57:41,066:DEBUG:4712445440:time_query: clearProcessedMessages: type: metadata.write_ts start: 1652465.999 end: 1652460.999
2022-05-13 10:57:41,440:DEBUG:4712445440:Found 0 messages in response to query {'user_id': UUID('4c689acb-16ea-4a84-bda0-5d5ab13cb87e'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652460.999, '$gte': 1652465.999}}
2022-05-13 10:57:41,440:DEBUG:4712445440:About to delete messages matching query {'user_id': UUID('4c689acb-16ea-4a84-bda0-5d5ab13cb87e'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652460.999, '$gte': 1652465.999}}
2022-05-13 10:57:41,440:DEBUG:4712445440:clearProcessedMessages: list: deleting 0 entries
2022-05-13 10:57:41,504:DEBUG:4712445440:clearProcessedMessages: count_docs: deleting 0 entries
```

I am looping on these failures, so we try again and succeed:
```
2022-05-13 10:57:42,688:DEBUG:4712445440:moveToLongTerm: Deleting all 30 entries for query <emission.storage.timeseries.timequery.TimeQuery object at 0x7fc380575350>
2022-05-13 10:57:42,688:DEBUG:4712445440:time_query: call clear with: type: metadata.write_ts start: 1652465.999 end: 1652460999
2022-05-13 10:57:42,688:DEBUG:4712445440:time_query: clearProcessedMessages: type: metadata.write_ts start: 1652465.999 end: 1652460999
2022-05-13 10:57:42,778:DEBUG:4712445440:Found 30 messages in response to query {'user_id': UUID('4c689acb-16ea-4a84-bda0-5d5ab13cb87e'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652460999, '$gte': 1652465.999}}
2022-05-13 10:57:42,778:DEBUG:4712445440:About to delete messages matching query {'user_id': UUID('4c689acb-16ea-4a84-bda0-5d5ab13cb87e'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652460999, '$gte': 1652465.999}}
2022-05-13 10:57:42,778:DEBUG:4712445440:clearProcessedMessages: list: deleting 30 entries
2022-05-13 10:57:42,848:DEBUG:4712445440:clearProcessedMessages: count_docs: deleting 30 entries
```
Notice end/lte changed from 1652460.999 to 1652460999 from one attempt to the next. The decimal disappeared. Also notice that for the first try,
start: 1652465.999  > end: 1652460.999.

**Comment by shankari** [comment]
Aha! the initial `write_ts` format seems wrong. It is supposed to be in seconds since the epoch, but `1652460.999` is way too small. Has the decimal point in the wrong place. concretely,

```
>>> arrow.get(1652460.999)
<Arrow [1970-01-20T03:01:00.999000+00:00]>
>>> arrow.get(1652460999)
<Arrow [2022-05-13T16:56:39+00:00]>
```

**Comment by aGuttman** [comment]
Yes! It seems that all successful deletes have lte as a 10 digit number and all failing have let as 7 digits with 3 decimals.

```
2022-05-13 10:59:00,731:DEBUG:4712445440:moveToLongTerm: Deleting all 30 entries for query <emission.storage.timeseries.timequery.TimeQuery object at 0x7fc33002d410>
2022-05-13 10:59:00,731:DEBUG:4712445440:time_query: call clear with: type: metadata.write_ts start: 1652466.117 end: 1652461.117
2022-05-13 10:59:00,731:DEBUG:4712445440:time_query: clearProcessedMessages: type: metadata.write_ts start: 1652466.117 end: 1652461.117
2022-05-13 10:59:00,799:DEBUG:4712445440:Found 0 messages in response to query {'user_id': UUID('1effebb2-472c-4004-aee2-898e52c64776'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652461.117, '$gte': 1652466.117}}
2022-05-13 10:59:00,799:DEBUG:4712445440:About to delete messages matching query {'user_id': UUID('1effebb2-472c-4004-aee2-898e52c64776'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652461.117, '$gte': 1652466.117}}
2022-05-13 10:59:00,799:DEBUG:4712445440:clearProcessedMessages: list: deleting 0 entries
2022-05-13 10:59:00,874:DEBUG:4712445440:clearProcessedMessages: count_docs: deleting 0 entries
```

The next attempt
```
2022-05-13 10:59:02,223:DEBUG:4712445440:moveToLongTerm: Deleting all 30 entries for query <emission.storage.timeseries.timequery.TimeQuery object at 0x7fc350180a90>
2022-05-13 10:59:02,223:DEBUG:4712445440:time_query: call clear with: type: metadata.write_ts start: 1652466.117 end: 1652461117.0
2022-05-13 10:59:02,223:DEBUG:4712445440:time_query: clearProcessedMessages: type: metadata.write_ts start: 1652466.117 end: 1652461117.0
2022-05-13 10:59:02,288:DEBUG:4712445440:Found 30 messages in response to query {'user_id': UUID('1effebb2-472c-4004-aee2-898e52c64776'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652461117.0, '$gte': 1652466.117}}
2022-05-13 10:59:02,288:DEBUG:4712445440:About to delete messages matching query {'user_id': UUID('1effebb2-472c-4004-aee2-898e52c64776'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}], 'metadata.write_ts': {'$lte': 1652461117.0, '$gte': 1652466.117}}
2022-05-13 10:59:02,288:DEBUG:4712445440:clearProcessedMessages: list: deleting 30 entries
2022-05-13 10:59:02,373:DEBUG:4712445440:clearProcessedMessages: count_docs: deleting 30 entries
```
I don't understand the workings of this whole thing. How is this changing run to run?

**Comment by shankari** [comment]
@aGuttman it almost seems like something assumes that this data is in milliseconds instead of seconds and is converting it to seconds by dividing by 1000. not sure how that can be documentDB though; it should theoretically not even know that this is a timestamp

**Comment by aGuttman** [comment]
Running local on mongo I can confirm that the numbers are always appropriately large. No dividing by 1000.

Looking at the time_query where moveToLongTerm first sees them format seems correct. Problem is coming from here then?
`unified_entry = enuf.convert_to_common_format(entry)`

Any of this look suspicious?
```
def convert_to_common_format(entry):
    format_fn = get_formatter(entry)
    return format_fn(entry)

def get_formatter(entry):
    module_name = get_module_name(entry.metadata) 
    logging.debug("module_name = %s" % module_name)
    module = importlib.import_module(module_name) 
    return getattr(module, "format")
   
def get_module_name(metadata):
    return "emission.net.usercache.formatters.{platform}.{key}".format(
        platform=metadata.platform,
        key=metadata.key.split("/")[1])
```

Did entry.metadata on the server get messed up somehow?

**Comment by shankari** [comment]
There are currently no entries in the usercache on DocumentDB, so the entries must be inserted as part of the test.

```
>>> import emission.core.get_database as edb
Connecting to database URL mongodb://...
>>> edb.get_usercache_db().count_documents({})
0
```

I do recall having some older entries in which the data was in milliseconds and have a special formatter for them. But then we should be using the same formatter on both mongodb and documentDB

**Comment by shankari** [comment]
Looking more closely at `moveToLongTerm`, the original time query is not based on the entries in the usercache.
Only on the entries in the pipeline state.

```
        time_query = esp.get_time_range_for_usercache(self.user_id)
```

```
def get_time_range_for_stage(user_id, stage):
...
        start_ts = curr_state.last_processed_ts
       end_ts = time.time() - END_FUZZ_AVOID_LTE
```

Let's check the start timestamps in the database, although it seems like `end_ts`, which uses `time.time()` should always be correct.

**Comment by shankari** [comment]
Start times in the pipeline state DB

```
[{'_id': ObjectId('627e8e26c20b9272a67e1c78'), .. 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652461122.0, 'last_ts_run': 1652461136.7204711}, {'_id': ObjectId('627e8e28c20b9272a67e1c79'), .. 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652461122, 'last_ts_run': 1652461138.545055}]
```

also seem to be correct

**Comment by shankari** [comment]
And we don't override the `start_ts` on the timequery anywhere in `moveToLongTerm`. So where do we get `metadata.write_ts start: 1652465.999` from?

It can't be from the DB because there's no entry with that number either in the pipeline state or in the usercache.

**Comment by shankari** [comment]
since it does work eventually, I wonder if the issue is that an earlier save (as part of the test) is messed up?

**Comment by shankari** [comment]
Here's a way to setup the database with all our entries. We can then run the test line by line and check the state of the database manually at each step.

```
$ ./e-mission-py.bash
Python 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 15:59:12)
[Clang 11.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import emission.core.get_database as edb
Connecting to database URL mongodb://
>>> list(edb.get_pipeline_state_db().find())
[{'_id': , 'user_id': , 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652477888.0, 'last_ts_run': 1652477905.4214709}, {'_id': , 'user_id': , 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652477888, 'last_ts_run': 1652477907.719307}]
>>> import emission.tests.netTests.TestBuiltinUserCacheHandlerInput as test
>>> test.TestBuiltinUserCacheHandlerInput().setUp()
collections = ['Stage_usercache', 'Stage_pipeline_state', 'Stage_timeseries']
Dropping collection Stage_usercache
Dropping collection Stage_pipeline_state
Dropping collection Stage_timeseries
>>> edb.get_usercache_db().count_documents({})
90
```

**Comment by shankari** [comment]
working on this a bit while I still have access...

**Comment by shankari** [comment]
Ran the `moveToLongTerm` 10 times

```
for i in 1 2 3 4 5 6 7 8 9
> do
> ./e-mission-py.bash emission/tests/netTests/TestBuiltinUserCacheHandlerInput.py TestBuiltinUserCacheHandlerInput.testMoveDuplicateKey > /tmp/test_run_$i 2>&1
> done
```

Results:

```
/tmp/test_run_1:AssertionError: 60 != 0
/tmp/test_run_2:AssertionError: 60 != 0
/tmp/test_run_3:AssertionError: 60 != 0
/tmp/test_run_4:AssertionError: 60 != 0
/tmp/test_run_5:AssertionError: 30 != 0
/tmp/test_run_6:AssertionError: 60 != 0
/tmp/test_run_7:no error
/tmp/test_run_8:AssertionError: 60 != 0
/tmp/test_run_9:no error
```

logs attached: 
[all_test_runs.tar.gz](https://github.com/e-mission/e-mission-docs/files/8695043/all_test_runs.tar.gz)

**Comment by shankari** [comment]
As already identified in
https://github.com/e-mission/e-mission-docs/issues/721#issuecomment-1126323085

in the failed run, we have timestamps that are divided by 1000

#### Failed run

```
2022-05-15 :INFO:For stage PipelineStages.USERCACHE, last_ts_processed = 1970-01-20T03:03:25.453000
2022-05-15 :DEBUG:About to save object PipelineState({'_id': ObjectId('6280aec57282e9a7d34ad72a'), 'user_id': UUID('6de1e5d4-6edd-4630-83da-cf7998bfd346'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652605.453, 'last_ts_run': 1652600512.310878})
2022-05-15 :DEBUG:After saving state PipelineState({'_id': ObjectId('6280aec57282e9a7d34ad72a'), 'user_id': UUID('6de1e5d4-6edd-4630-83da-cf7998bfd346'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652605.453, 'last_ts_run': 1652600512.310878}), list is [{'_id': ObjectId('6280aec57282e9a7d34ad72a'), 'user_id': UUID('6de1e5d4-6edd-4630-83da-cf7998bfd346'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652605.453, 'last_ts_run': 1652600512.310878}]
2022-05-15 :DEBUG:Found 60 messages in response to query {'user_id': UUID('6de1e5d4-6edd-4630-83da-cf7998bfd346'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}]}
```

#### Successful run

```
2022-05-15 :INFO:For stage PipelineStages.USERCACHE, last_ts_processed = 2022-05-15T07:41:48
2022-05-15 :DEBUG:About to save object PipelineState({'_id': ObjectId('6280aefaa6a91fb9be3f353f'), 'user_id': UUID('cf55a402-fede-4c94-87c4-f012a5049ff5'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652600508, 'last_ts_run': 1652600565.0036852})
2022-05-15 :DEBUG:After saving state PipelineState({'_id': ObjectId('6280aefaa6a91fb9be3f353f'), 'user_id': UUID('cf55a402-fede-4c94-87c4-f012a5049ff5'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652600508, 'last_ts_run': 1652600565.0036852}), list is [{'_id': ObjectId('6280aefaa6a91fb9be3f353f'), 'user_id': UUID('cf55a402-fede-4c94-87c4-f012a5049ff5'), 'pipeline_stage': 0, 'curr_run_ts': None, 'last_processed_ts': 1652600508, 'last_ts_run': 1652600565.0036852}]
2022-05-15 :DEBUG:Found 0 messages in response to query {'user_id': UUID('cf55a402-fede-4c94-87c4-f012a5049ff5'), '$or': [{'metadata.type': 'message'}, {'metadata.type': 'sensor-data'}, {'metadata.type': 'rw-document'}]}
2022-05-15 :DEBUG:curr_query = {'invalid': {'$exists': False}, 'user_id': UUID('cf55a402-fede-4c94-87c4-f012a5049ff5')}, sort_key = metadata.write_ts
```

**Comment by shankari** [comment]
hmm, there's some weirdness here.

1. the write_ts for the entries seems to differ between runs. Need to see whether this is intentional in the test.

2. the write_ts in both runs is frequently off

```
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.activity
2022-05-15 :DEBUG:write_ts = 1652600183
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.location
2022-05-15 :DEBUG:write_ts = 1652600.183
```

In fact, the location timestamps are consistently off, even in the successful run

```
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.location
2022-05-15 :DEBUG:write_ts = 1652600.384
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.location
2022-05-15 :DEBUG:write_ts = 1652600.413
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.location
2022-05-15 :DEBUG:write_ts = 1652600.443
2022-05-15 :DEBUG:module_name = emission.net.usercache.formatters.android.location
2022-05-15 :DEBUG:write_ts = 1652600.473
```

**Comment by shankari** [comment]
To answer my own questions, yes, we do set the write_ts in the test

```
        self.curr_ts = int(time.time())
        for offset in range(self.curr_ts - 5 * 60, self.curr_ts, 30):
            for entry in self.entry_list:
                entry["metadata"]["write_ts"] = offset
            mauc.sync_phone_to_server(self.testUserUUID1, self.entry_list)
```

But when they are written, the `write_ts` is fine

```
2022-05-15 :DEBUG:Updated result for user = 7d10ca56-d7a6-4342-9729-0de94d04ad56, key = background/activity, write_ts = 1652600384 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aee9aa2dde671008178a'), 'ok': 1.0, 'operationTime': Timestamp(1652600553, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = 7d10ca56-d7a6-4342-9729-0de94d04ad56, key = background/location, write_ts = 1652600384 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aee9aa2dde671008178b'), 'ok': 1.0, 'operationTime': Timestamp(1652600553, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = 7d10ca56-d7a6-4342-9729-0de94d04ad56, key = statemachine/transition, write_ts = 1652600384 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aeeaaa2dde6540464ff1'), 'ok': 1.0, 'operationTime': Timestamp(1652600554, 1), 'updatedExisting': False}
```

It is also interesting that this is the last entry for user 7d10ca56-d7a6-4342-9729-0de94d04ad56

```
2022-05-15 :DEBUG:Updated result for user = 7d10ca56-d7a6-4342-9729-0de94d04ad56, key = statemachine/transition, write_ts = 1652600384 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aeeaaa2dde6540464ff1'), 'ok': 1.0, 'operationTime': Timestamp(1652600554, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/motion_activity, write_ts = 1652600233 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aeeaaa2dde6540464ff2'), 'ok': 1.0, 'operationTime': Timestamp(1652600554, 1), 'updatedExisting': False}
```

so where are the other timestamps in https://github.com/e-mission/e-mission-docs/issues/721#issuecomment-1126952817 (1652600.413, 1652600.443) coming from?

They are coming from the next user.

```
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/motion_activity, write_ts = 1652600413 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef0aa2dde6540465004'), 'ok': 1.0, 'operationTime': Timestamp(1652600560, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/location, write_ts = 1652600413 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef0aa2dde6540465005'), 'ok': 1.0, 'operationTime': Timestamp(1652600560, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = statemachine/transition, write_ts = 1652600413 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef0aa2dde6540465006'), 'ok': 1.0, 'operationTime': Timestamp(1652600560, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/motion_activity, write_ts = 1652600443 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef1aa2dde6540465007'), 'ok': 1.0, 'operationTime': Timestamp(1652600561, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/location, write_ts = 1652600443 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef1aa2dde6540465008'), 'ok': 1.0, 'operationTime': Timestamp(1652600561, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = statemachine/transition, write_ts = 1652600443 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aef1aa2dde6540465009'), 'ok': 1.0, 'operationTime': Timestamp(1652600561, 1), 'updatedExisting': False}
```

But they are not the first entries for that user

```
2022-05-15 :DEBUG:Updated result for user = 7d10ca56-d7a6-4342-9729-0de94d04ad56, key = statemachine/transition, write_ts = 1652600384 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aeeaaa2dde6540464ff1'), 'ok': 1.0, 'operationTime': Timestamp(1652600554, 1), 'updatedExisting': False}
2022-05-15 :DEBUG:Updated result for user = b127c844-3fd0-44e9-a6ce-1b1d471fd382, key = background/motion_activity, write_ts = 1652600233 = {'n': 1, 'nModified': 0, 'upserted': ObjectId('6280aeeaaa2dde6540464ff2'), 'ok': 1.0, 'operationTime': Timestamp(1652600554, 1), 'updatedExisting': False}
```

So why are we seeing these out of order?

**Comment by shankari** [comment]
Ok so I know where we are doing the conversion. if we have "old-style" android messages, we assume that the timestamps are in milliseconds

```
    if ("mLatitude" in entry.data):
        return format_location_raw(entry)
    else:
        return format_location_simple(entry)
```

and 

```
    metadata.time_zone = "America/Los_Angeles"
    metadata.write_ts = old_div(float(entry.metadata.write_ts), 1000)
```

but we do that for both the working and not working case, so that is not the difference between them.

**Comment by shankari** [comment]
Re-ran with additional logging. Results this time were:

```
grep AssertionError *
test_run_more_logs_1:AssertionError: 60 != 0
test_run_more_logs_2:AssertionError: 30 != 0
test_run_more_logs_3:AssertionError: 30 != 0
test_run_more_logs_4:AssertionError: 60 != 0
test_run_more_logs_5:AssertionError: 60 != 0
test_run_more_logs_6:AssertionError: 60 != 0
test_run_more_logs_7:AssertionError: 60 != 0
test_run_more_logs_8:AssertionError: 60 != 0
test_run_more_logs_9:AssertionError: 60 != 0
```

Let's start by comparing run 1 and run 2

**Comment by shankari** [comment]
So the new logs don't actually have the additional logs that I thought I added, but I was able to use them anyway.

The difference seems to happen when we have multiple entries with the same `write_ts`. In that case, it looks like MongoDB returns them in a consistent order, but document DB returns them in a different order.

For example, consider runs 1 and 2 above.

For the first set of entries processed (`write_ts = 1652644367` on the first run and `1652644404` on the second run), we read entries in the order (activity, location, transition) for both runs.

<img width="1382" alt="Screen Shot 2022-05-16 at 10 03 04 PM" src="https://user-images.githubusercontent.com/2423263/168732977-411d2287-312c-4f54-89ec-3de4a4db60b3.png">

For write_ts = `1652644367` and `1652644434` right below, we read entries in the order (activity, location, transition) in one and in the order (transition, activity, location) in the second.

<img width="1382" alt="Screen Shot 2022-05-16 at 9 58 21 PM" src="https://user-images.githubusercontent.com/2423263/168732499-87aaeffb-d349-42d7-ab17-11518cbbd741.png">

and for the last set of entries processed before the tests diverge, the order on the second run is with the location last.

<img width="1382" alt="Screen Shot 2022-05-16 at 10 12 36 PM" src="https://user-images.githubusercontent.com/2423263/168733934-c7edb51a-984d-41c3-897d-1cebda40d6f9.png">

 So the last timestamp is the incorrect one and the we don't delete anything and the test fails.

<img width="1382" alt="Screen Shot 2022-05-16 at 10 13 38 PM" src="https://user-images.githubusercontent.com/2423263/168734052-72373777-a25c-4b74-a05e-bf700dcb806c.png">

@aGuttman not sure what the order should be on MongoDB vs documentDB if the sort key is the same.

**Comment by shankari** [comment]
And in the final failure on run 1 also, location is last

```
2022-05-15 12:57:52,107:DEBUG:4475928064:module_name = emission.net.usercache.formatters.android.activity
2022-05-15 12:57:52,107:DEBUG:4475928064:write_ts = 1652644607
2022-05-15 12:57:52,107:DEBUG:4475928064:module_name = emission.net.usercache.formatters.android.transition
2022-05-15 12:57:52,107:DEBUG:4475928064:Timestamp conversion: 1652644607 -> 1652644607 done
2022-05-15 12:57:52,107:DEBUG:4475928064:write_ts = 1652644607
2022-05-15 12:57:52,108:DEBUG:4475928064:Mapped local.state.waiting_for_trip_start -> 1
2022-05-15 12:57:52,108:DEBUG:4475928064:No existing timestamp, copyied from metadata1652644607
2022-05-15 12:57:52,108:DEBUG:4475928064:module_name = emission.net.usercache.formatters.android.location
2022-05-15 12:57:52,108:DEBUG:4475928064:write_ts = 1652644.607
```

**Comment by aGuttman** [comment]
I should have checked back here earlier, I just found the same thing myself and was excited to report back.

I added an extra log to `format_location_raw()` and it looks like it always creates a number that is unreasonably small. There doesn't seem to be a legitimate example where dividing by 1000 creates a time stamp that we want. Can we just get rid of the division? Then the ordering shouldn't matter.

I guess to add to that, I never see `format_location_simple()` being used. Perhaps `if ("mLatitude" in entry.data):` is branching off the wrong condition?

I think that's the logical root of this. All entries have a mLatitude. `entry.data["mLatitude"]` is  `37.3885529` for all of the entries being tested which makes it seem like some sort of default value. I would look into it further right now, but I just spent 5 minutes getting really frustrated with grep before realizing I was in a completely wrong directory, so I think I'm at my limit for the day.

**Comment by aGuttman** [comment]
Ok, I didn't really understand how the test examples were being generated before but I see now.

All entries are based off of those in `emission/tests/data/netTests/`. `android.location.raw.txt` from there is an "old style" entry with an mLatitude field and a write_ts in milliseconds (13 digits). However, when we populate the user cache with entries, we are overwriting write_ts with times based on the current time, in seconds (10 digits). So we are giving these old style android location entries a write_ts in the wrong format, which was always a problem, regardless of what database we use, but the problem was hidden on mongoDB since the entries always came in (activity, location, transition) order, so the location entires were never last and thus their write_ts were never used in the query to `clearProcessedMessages()`.

While the ordering being consistent on mongo and not on document is notable and good to know going forward, the way location entries were formatted was already wrong.

**Comment by aGuttman** [comment]
I tried giving location entries a write_ts in milliseconds in `setUp()` to see what would happen, thinking it would probably break `moveToLongTerm()`'s query, but it didn't. Turns out `sync_phone_to_server()` in `emission/net/api/usercache.py` already checks for timestamps in milliseconds and converts them.
```
if ecc.isMillisecs(data["metadata"]["write_ts"]):
        data["metadata"]["write_ts"] = old_div(float(data["metadata"]["write_ts"]), 1000)

if "ts" in data["data"] and ecc.isMillisecs(data["data"]["ts"]):
        data["data"]["ts"] = old_div(float(data["data"]["ts"]), 1000)
```

Given this, is there any case where timestamps in the user cache can be in milliseconds? What over functions can put data in the user cache?

Is `format_location_raw()` from 'emission/net/usercache/formatters/android/location.py' ever used on entries not in the user cache? If no, we can remove the division from it. If yes, we need another function or check inside of the existing function for entries that have already been converted. Otherwise we divide timestamps for old style entires by 1000*1000 instead of 1000.

**Comment by aGuttman** [comment]
Added server name confining in URL and hack around millisecond division commits to my pull request.

**Comment by shankari** [comment]
@aGuttman I understand the urge to fix the test, and yes, we can remove the hacks/workarounds finally now that we don't use "old style" formats any more.

But wrt the meeting that we are going to attend in a few hours with @jgu2, the issue is another inconsistency between mongodb and documentDB

In other words:
- how many other places in the code does this inconsistency affect?
- how many other such inconsistencies between mongo and documentDB exist?

It's great that we have a test framework and automated tests, but we don't currently have a defensible estimate of the code coverage for the tests.

**Comment by aGuttman** [comment]
Understood. Another hit against the use of documentDB is lack of online resources covering it. There is the official amazon documentation and very little else. If I don't understand something from the documentation or suspect it might be inaccurate, there are very often no discussions about it.

I looked for information about the ordering inconsistency and didn't find anything.

**Comment by shankari** [comment]
@aGuttman ok, let's see what the cloud services (@jgu2) team says about a mongodb cluster.

**Comment by aGuttman** [comment]
>The way that location entries were formatted was not wrong, it was right for that obsolete code branch that ended up being inconsistent with that particular test case.

I should have been more specific. The test inserts a timestamp in seconds into old style entries which should be in milliseconds, and when the old style formatter is applied to them, it misformats them since it was expecting milliseconds. Inserting the seconds timestamp into the old style entires in this test was already creating misformatted entries, regardless of database.

**Comment by shankari** [comment]
>  Inserting the seconds timestamp into the old style entires in this test was already creating misformatted entries, regardless of database.

Agreed. But the "incorrect format of the entries in the database" are:
- something we would not encounter in the real world any more
- something that the test doesn't really care about since we don't explicitly check for it

What we care about is that the correct number of entries is moved to long term

Maybe we should care about the format of the entries as well and not just their being copied correctly, but that is part of the code coverage discussion...

**Comment by shankari** [comment]
@aGuttman I see that you have included the fix for the incorrectly formatted ts in your PR, but I'd rather just remove this obsolete code path than hack around it some more.

I originally included the hack in 2015 (https://github.com/e-mission/e-mission-server/commit/8cab86f867dc32f8543f55ac28efe9e5c082b098) to provide backwards compatibility and a consistent migration path while changing the location format, since the phone clients can take a while to be upgraded.

It is now almost 5 years later. We are working on publishing a brand new client. Keeping obsolete code paths around increases complexity unnecessarily, and (as we saw) can make the codebase harder to reason about.

Let's remove this codepath completely. @aGuttman I know it can be scary to remove code that you haven't written, but creative destruction is an important part of keeping code manageable. LMK if you would like me to make the change instead this first time.

**Comment by shankari** [comment]
wrt "Maybe we should care about the format of the entries as well", I checked and we do have a test explicitly for that 

https://github.com/e-mission/e-mission-server/blame/2177582fdea9a52faa8c1a238e62b111dbbcedba/emission/tests/netTests/TestFormatters.py#L39

Of course, in this case, the write_ts is not overridden so it still works.

**Comment by aGuttman** [comment]
I did some more looking around while on the call and did find documentation of sort ordering issues:
https://docs.aws.amazon.com/documentdb/latest/developerguide/functional-differences.html#functional-differences.result-ordering
If I understand it correctly, this confirms that our problem is known, sorts will be ordered based on what they are given, but nothing else is guaranteed. Also if something is sorted by one field and then sorted by another field, mongo will preserve the first ordering in the second, where documentDB does not.

I don't know if this occurs and is important anywhere else in our code, but if it is, doing a compound sort could fix it. I added a compound sort to `getMessage()`
```
retrievedMsgs = list(self.db.find(combo_query).sort([("metadata.write_ts", pymongo.ASCENDING),("metadata.key", pymongo.ASCENDING)]).limit(edb.result_limit))
```
Which prevents locations from coming last and it works. Adding this and removing my hack fix allows the tests to pass.

**Comment by shankari** [comment]
@aGuttman @jgu2 We are looking into a couple of options:

1. using mongodb in a container as a transition while we experiment further with documentDB
  - this will require additional cyber approval and may take more time
  - it is only a stopgap
2. go ahead with documentDB
  - get the current tests to pass by removing the code path for the "old style" locations
  - figure out a way to ask our questions to an AWS engineer
  - figure out a way to run tests in the AWS environment
  - determine the code coverage of the existing codebase and potentially bulk up on tests to ensure that we don't have other gremlins lurking in the code

tracking integration with a code coverage tool at https://github.com/e-mission/e-mission-docs/issues/729

**Comment by aGuttman** [comment]
I missed something important.

```
Traceback (most recent call last):
  File "emission/tests/storageTests/TestMongoGeoJSONQueries.py", line 35, in testGeoWithinPostsData
    for a in self.Sections.find({ "loc" : { "$geoWithin" : { "$polygon" :[ [ 90,31 ],[90,40] ,[ 110,40 ],[110,31]] } } }):

...

pymongo.errors.OperationFailure: unknown geo specifier: $polygon, full error: {'ok': 0.0, 'operationTime': Timestamp(1653342214, 1), 'code': 2, 'errmsg': 'unknown geo specifier: $polygon'}
```

When I ran the unit tests on remote I had the problem of my computer going to sleep serval times. These seemed to produce some timeout errors, which I thought weren't a worry, because when I went back and reran tests without sleeping/individually, they seemed to go away.

I must have overlooked this one somehow.

From https://docs.aws.amazon.com/documentdb/latest/developerguide/geospatial.html:

> Amazon DocumentDB does not support querying or indexing of Polygons, LineString, MultiPoint, MultiPolygon, MultiLineString, and GeometryCollection.

Which is odd, because earlier on the same page:

> A good example of inclusion querying is to find all points (all airports) that located in a specified area/polygon (state of New York). A good example of intersection query is finding a polygon (state) which intersects with a point (city). You can use the following Geospatial operators to gain insights from your data.
>...
> * $geoWithin - $geoWithin is a find operator that supports finding documents with geospatial data that exists entirely within a specified shape such as a polygon.

This, of course, all works fine on my local mongodb.

**Comment by shankari** [comment]
@aGuttman great catch! We should definitely add this to the list of DocumentDB compat issues to flag.

Fortunately, IIRC, we are not actively using the `$geowithin` functionality actively right now.

To think this through:
- I used to use it for heatmaps, but we won't have the deployer dashboard up immediately.
- I also use it for targeted surveys (people who typically travel a route), but again, we can defer that until we actually have some data to survey people about
- I will want to use it to automatically split the open access data into geographic regions in the future but that is in the future

**Comment by shankari** [comment]
@aGuttman I see that  you have made the changes for https://github.com/e-mission/e-mission-server/pull/849
I will merge it as soon as the tests finish.

Can you run the tests against DocumentDB from the newly created test AWS instance (see instructions from @jgu2) and see if there are any additional timing issues that we encounter in the AWS <-> AWS environment?

**Comment by shankari** [comment]
Now that https://github.com/e-mission/e-mission-server/pull/849 has been merged and https://github.com/e-mission/e-mission-docs/issues/727 has been resolved, we are ready to deploy the NREL OpenPATH instance to staging.

Let's track that at https://github.com/e-mission/e-mission-docs/issues/732

**Comment by aGuttman** [comment]
## Running unit tests on AWS ec2 instance with DocumentDB:
- runAllTests.sh runs in about 40 minutes, as opposed to ~15 when running locally with mongoDB
  - Generates 15 errors
    - 5 errors due to use of polygon queries
      - Issue discussed above
    - 10 errors due to auth error on pymongo's drop_database()
      - From David Rager: "FYI – on shared clusters we do not allow create/drop database permissions to protect the assets of other projects on cluster."
  - Other than these expected errors, passes all other tests

## Database Name
Database names are configurable through conf/storage/db.conf. Use the desired name after `:27017/` in the url.
ie:
`mongodb://<inserYourUsername>:<insertYourPassword>@documentdb-moderate-1.cluster-ccklrxkcenui.us-west-2.docdb.amazonaws.com:27017/<db_name>?<...>`

Note that cloud services will still need to create the database with the desired name and give the user access to it. This only accesses existing DBs that have been set up with the proper permissions, it does not create them.

## Read Pref on DocumentDB
- Setting read preference through database url seems to be the cleanest solution
  - Put `&readPreference=primaryPreferred` in url ie: `mongodb://<inserYourUsername>:<insertYourPassword>@documentdb-moderate-1.cluster-ccklrxkcenui.us-west-2.docdb.amazonaws.com:27017/<db_name>?ssl=true&ssl_ca_certs=rds-combined-ca-bundle.pem&replicaSet=rs0&readPreference=primaryPreferred&retryWrites=false`
- Solves the read-after-write issue documented above
  - The read-after-write test fails when tested on ec2 with `readPreference=secondaryPreferred` and no sleep. It succeeds with `readPreference=primaryPreferred` and no sleep. When tested with server run locally connecting to remote DocumentDB, network creates enough delay that never fails, regardless of read pref.

## Code Coverage on DocumentDB
Code coverage drops for 72% with MongoDB to 70% on DocumentDB. The following files make up this drop:
- analysis/classification/inference/mode/seed/pipeline.py
  - 90% to 21%
- emission/analysis/classification/inference/mode/seed/section_features.py
  - 38% to 13%
- core/common.py
  - 82% to 69%
- emission/core/get_database.py
  - 70% to 68%
- emission/tests/analysisTests/modeinferTests/TestPipelineSeed.py
  - 98% to 27%
- emission/tests/storageTests/TestMongoGeoJSONQueries.py
  - 90% to 68%

Looking at this quickly, it appears that these drops are almost all due to the errors mentioned above causing tests to quit early (ie: pipeline.py has functions that get called in /emission/tests/analysisTests/modeinferTests/TestPipelineSeed.py, which has a drop_database() early on in a few tests, so it never calls those functions.) The differences between MongoDB and DocumentDB are not so large that they cause different branches to fire between the test cases.

## Encountered DocumentDB Incompatibilities
- Hashed indices
  - Not supported. Switched hashed to ascending. Makes no noticeable time difference in local MongoDB test.
- Geosphere indices
  - Supported, but cannot be sparse. Removed sparse, makes no noticeable time difference in local MongoDB tests.
- Stable sorting
  - MongoDB sorts are stable and consistent. DocumentDB sorts are not. Best thing to do is to avoid design decisions that rely on stable sorting, but if necessary, compound sorts can be used to create similar (but not exactly that same) results*.
- Polygon quieres/indices
  - Not supported. No easy workaround found.
- Replica Sets/Read Preference
  - read-after-write issue documented above. Solved with `readPreference=secondaryPreferred` in url.


<details>
<summary>*Sorting Details (click to unfold)</summary>
MongoDB sorts return the same result over the same collection of items every time. That is, ties will always be broken the same way; whatever was first in the original collection will be first in the result. This is known as stable sorting. DocumentDB does not guarantee this. Sorts will not be incorrect, but could be different run to run where items that are tied might switch places. Ideally, things would be designed in a way where this is not a problem, but stable sorting can hide assumptions. Say we sort Collection by Index1, ascending. Later we sort Collection by Index2, ascending. After this second sort, we might unknowingly rely on the fact that no items of lower Index1 value can follow an item of higher Index1 value when they all have have equal Index2 values. Making this assumption in DocumentDB will cause errors. We can fix this by doing a compound sort by (Index2, ascending, Index1, ascending), but we must notice that this is happening first, which is tricky to see. Even with a compound sort we are not guaranteed exactly the same result in DocumentDB that we got in MongoDB, because ties are still broken "randomly" when they exist, (ie: Index2 and Index1 are equal.) We just have a second (or more) rule to fall back to after the first, but orderings can still change run to run. That said, if this causes problems, the code is very fragile and the design should be looked at anyway.
</details>
